lib/executor.js

```diff
@@ -1,57 +1,103 @@
-    measureIdsList.forEach((id) => {
-      const mes = this.measureSource.getMeasure(id);
-      measures.push(mes.generate(options));
+    measureIdsList.forEach(async (id) => {
+      const mes = this.measureSource.findMeasure(id);
+      measurePromises.push(mes);
     });
-    return measures;
+    return Promise.all(measurePromises);
+  }
+
+  /*
+  Generates the QDM patients for calculation
+  @params [array] patientIds -- the patients to generate
+  @returns [Promise] a promise of executable patients
+
+  */
+  async getPatientsById(patientIds) {
+    // patientIds parameter will either be a single id or multiple ids.
+    // If the former, transform it to an array
+    const patientIdsList = Array.isArray(patientIds) ? patientIds : [patientIds];
+
+    return this.patientSource.findPatients(patientIdsList);
   }
 
   /*
    Executes measures against a given patient source
-   @params [PatientSource] patientSource -- the source for patient records.
+   @params {array} patientIds -- the array of patient ids for which results are requested. The patients all adhere to the QDM data model in MongoDB
 
    @params {array} measureIds -- the array of measures to execute. The ids can be CMS ids with the included sub_id if applicable, the hqmfId with the
    included sub_id if applicable or a json object that contains {measure_id: (CMS or HQMF_ID), sub_id: "" }
 
-   @params {object} handler -- the pateint level results handler.  During calculation patient level results will be given to this object to handle in some fashion
+   @params {object} options -- json object with calculation option such as handler information, etc ....
+
+   @params {object} options - handler information -- the pateint level results handler.  During calculation patient level results will be given to this object to handle in some fashion
    such as saving to a database.  The handler needs to implement the start() finsihed() and handleResult(p_level_result) methods.  start() should basically reset the handler and
    get it ready for calculation, handleResult(presults) will perfrom some task on the handler and finished() allows the handler to know when the paitent
    level calculations are finsihed and it can perform any post processing that needs to be perfromed, such as aggregating results
 
-   @params {object} options -- json object with calculation option ssuch as effective_date , enable_logging ....
    */
-  execute(patientSource, measureIds, handler, options) {
-    handler.start();
-    const measures = this.generate_measures(measureIds, options);
+  async execute(patientIds, measureIds, options = null) {
+    await this.getPatientsById(patientIds);
+    const measures = await this.getMeasuresById(measureIds);
+
+    const executor = this;
+
+    let handler;
+
+    if (!options || !options.handlerType || options.handlerType === 'CypressMongo') {
+      handler = new CypressHandler(this.connectionInfo);
```

Comment:
Is there a way to make this more flexible to better handle different "HandlerTypes"?